### PR TITLE
issue: 1065079 Add Memic support

### PR DESCRIPTION
**Pull request by liranoz12**
Use the internal chip memory (ICMC) while data exceed BF size.

Signed-off-by: Liran Oz <lirano@mellanox.com

### CODE REVIEW

**Comment by liranoz12** [comment]
bot:retest:MellanoxLab-RedmineChecker

**Comment by liranoz12** [comment]
bot:retest

File: configure.ac [hunk 1]
```diff
@@ -218,6 +218,23 @@ AS_IF([test "x$enable_exp_cq" == xyes],
 	])
 )
 
+# Device Memory
```
Comment (rafiw) [inline configure.ac]: this is shorter
AC_CHECK_DECL([IBV_EXP_DEVICE_ATTR_MAX_DM_SIZE,
		[AC_DEFINE(HAVE_IBV_DM, 1, [Define to 1 if IBV_EXP_DEVICE_ATTR_MAX_DM_SIZE is defined])],
		[],
		[[#include <infiniband/verbs_exp.h>]])

File: src/vma/dev/qp_mgr_eth_mlx5.h [hunk 2]
```diff
@@ -58,15 +60,17 @@ friend class cq_mgr_mlx5;
 private:
 	cq_mgr*		init_rx_cq_mgr(struct ibv_comp_channel* p_rx_comp_event_channel);
 	virtual cq_mgr* init_tx_cq_mgr(void);
+	virtual bool    is_compilation_need() { return !m_n_unsignaled_count || m_dm_context.dm_request_completion(); };
```
Comment (rafiw) [inline src/vma/dev/qp_mgr_eth_mlx5.h]: compilation -> completion

File: src/vma/util/dm_context.cpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+		ibv_exp_free_dm(ibv_dm);
+		dmc_logerr("ibv_exp_free_dm error - dm_mr registration failed, %d %m", errno);
+		return false;
+	}
+
+	m_allocation = allocation_size;
+	m_p_mlx5_dm = reinterpret_cast<struct vma_mlx5_dm *> (ibv_dm);
+	m_p_ring_stat->n_tx_dev_mem_allocated = m_allocation;
+
+	dmc_logwarn("Device memory allocation completed successfully! device[%s] bytes[%zu] dm_mr handle[%d] dm_mr lkey[%d]",
```
Comment (rafiw) [inline src/vma/util/dm_context.cpp]: why warning?

File: configure.ac (hunk 1 above)
Comment (liranoz12) [reply configure.ac]: +1

File: src/vma/dev/qp_mgr_eth_mlx5.h (hunk 2 above)
Comment (liranoz12) [reply src/vma/dev/qp_mgr_eth_mlx5.h]: :+1:

File: src/vma/util/dm_context.cpp (hunk 3 above)
Comment (liranoz12) [reply src/vma/util/dm_context.cpp]: for debug reasons. of course it will not be merged as is now.

**Comment by ilansmith** [comment]
:+1:

File: src/vma/util/dm_context.cpp [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+		return false;
+	}
+
+	if (!ib_ctx->get_on_device_memory_size()) {
+		// On Device Memory usage is not supported
+		return false;
+	}
+
+	// Initialize attributes
+	bzero(&mr_attr, sizeof(mr_attr));
```
Comment (rafiw) [inline src/vma/util/dm_context.cpp]: this is depreceted use memset

**Comment by liranoz12** [comment]
Merged #462

**Pull request closed without merging**